utilities/cache_dump_load_impl.cc

```diff
@@ -105,77 +105,57 @@ bool CacheDumperImpl::ShouldFilterOut(const Slice& key) {
+  return [&](const Slice& key, Cache::ValueType* value, size_t /*charge*/,
+             const Cache::CacheItemHelper* helper) {
+    if (helper == nullptr || helper->size_cb == nullptr ||
+        helper->saveto_cb == nullptr) {
+      // Not compatible with dumping. Skip this entry.
+      return;
     }
-    bool filter_out = false;
 
-    // Step 2: based on the key prefix, check if the block should be filter out.
-    if (ShouldFilterOut(key)) {
-      filter_out = true;
-    }
+    CacheEntryRole role = helper->role;
+    CacheDumpUnitType type = CacheDumpUnitType::kBlockTypeMax;
 
-    // Step 3: based on the block type, get the block raw pointer and length.
-    const char* block_start = nullptr;
-    size_t block_len = 0;
     switch (role) {
       case CacheEntryRole::kDataBlock:
         type = CacheDumpUnitType::kData;
-        block_start = (static_cast<Block*>(value))->data();
-        block_len = (static_cast<Block*>(value))->size();
         break;
       case CacheEntryRole::kFilterBlock:
         type = CacheDumpUnitType::kFilter;
-        block_start = (static_cast<ParsedFullFilterBlock*>(value))
-                          ->GetBlockContentsData()
-                          .data();
-        block_len = (static_cast<ParsedFullFilterBlock*>(value))
-                        ->GetBlockContentsData()
-                        .size();
         break;
       case CacheEntryRole::kFilterMetaBlock:
         type = CacheDumpUnitType::kFilterMetaBlock;
-        block_start = (static_cast<Block*>(value))->data();
-        block_len = (static_cast<Block*>(value))->size();
         break;
       case CacheEntryRole::kIndexBlock:
         type = CacheDumpUnitType::kIndex;
-        block_start = (static_cast<Block*>(value))->data();
-        block_len = (static_cast<Block*>(value))->size();
-        break;
-      case CacheEntryRole::kDeprecatedFilterBlock:
-        // Obsolete
-        filter_out = true;
-        break;
-      case CacheEntryRole::kMisc:
-        filter_out = true;
-        break;
-      case CacheEntryRole::kOtherBlock:
-        filter_out = true;
-        break;
-      case CacheEntryRole::kWriteBuffer:
-        filter_out = true;
         break;
       default:
-        filter_out = true;
+        // Filter out other entries
+        // FIXME? Do we need the CacheDumpUnitTypes? UncompressionDict?
+        return;
     }
 
-    // Step 4: if the block should not be filter out, write the block to the
-    // CacheDumpWriter
-    if (!filter_out && block_start != nullptr) {
-      WriteBlock(type, key, Slice(block_start, block_len))
-          .PermitUncheckedError();
+    // based on the key prefix, check if the block should be filter out.
+    if (ShouldFilterOut(key)) {
+      return;
+    }
+
+    assert(type != CacheDumpUnitType::kBlockTypeMax);
+
+    // Use cache item helper to get persistable data
+    // FIXME: reduce copying
+    size_t len = helper->size_cb(value);
+    buf.assign(len, '\0');
```

Comment:
Is this necessary, since the saveto callback will overwrite it anyway?

Reply:
The only way to allocate space in a std::string that is usable as a char buffer is to initialize it.